front/scripts/i18n-checker.ts

```
@@ -0,0 +1,145 @@
+  const tempDir = await mkdtemp(`${os.tmpdir()}/osrd-i18n-`);
+  return new Promise((resolve, reject) => {
+    const stream = vfs
+      .src(`${process.cwd()}/src/**/*.{ts,tsx}`)
+      .pipe(
+        // eslint-disable-next-line
+        new (I18nextParser as any).gulp({
+          locales: ['fr'],
+          output: '$LOCALE/$NAMESPACE.json',
+        })
+      )
+      .pipe(vfs.dest(tempDir));
+
+    stream.on('finish', () => resolve(tempDir));
+    stream.on('error', (e: Error) => reject(e));
+  });
+}
+
+/**
+ * The script execution
+ */
+async function run() {
+  try {
+    const appNamespacedKeys = await getLocalesKeys(`${process.cwd()}/public/locales`);
+
+    const scannedLocalePath = await scanCode();
+    const scannedNamespacedKeys = await getLocalesKeys(scannedLocalePath);
+
+    // Search for unused keys
+    const unusedKeys: string[] = [];
+    appNamespacedKeys.forEach((key) => {
+      if (
+        !scannedNamespacedKeys.has(key) &&
+        IGNORE_UNUSED.every((pattern) => !key.match(pattern))
+      ) {
+        unusedKeys.push(key);
+      }
+    });
+
+    // Search for missing traduction
+    const missingKeys: string[] = [];
+    scannedNamespacedKeys.forEach((keyWithNS) => {
+      if (
+        !appNamespacedKeys.has(keyWithNS) &&
+        IGNORE_MISSING.every((pattern) => !keyWithNS.match(pattern))
+      ) {
+        missingKeys.push(keyWithNS);
+      }
+    });
+
+    /* eslint-disable no-console */
+    if (unusedKeys.length > 0) {
+      console.log(`Unused keys (${unusedKeys.length})`);
+      console.log('----------------------------------');
+      console.log(unusedKeys.join('\n'));
+      console.log();
+    }
+
+    if (missingKeys.length > 0) {
+      console.log(`Missing keys (${missingKeys.length})`);
+      console.log('------------------------------------');
+      console.log(missingKeys.join('\n'));
+      console.log();
+      console.log('/!\\ Failed: missing keys are not allowed');
+      process.exit(1);
```

We should throw an error if there is at least 1 unused or missing key
```suggestion
  }
  if (unusedKeys.length > 0 || missingKeys > 0) {
      process.exit(1);
```

---

"unused" means that the script that parse the  code has not found the key. This can happened if the i18n key is dynamically generated (like  ` t(`error.${error.code}.message`)`). 

And we have many like that, even if there is a `IGNORE_UNUSED`  pattern list.

It will be better to ask for dev to improve this list, but I think it will painfull.

---

Okay, got it, thanks 🙏